Added firmware 2.8.0 support

Chris971991 opened the pull request:
- Added firmware 2.8.0 support 
- Automatic network testing functions to scan the network and test all functions of the air conditioners (test\ python integrated_network_test.py)

The idea is to have this update integrated into the official Daikin Home Assistant integration to support firmware 2.8.0

fredrike commented:
Thank you for this long awaited PR. I've not seen it before (must adjust my notifications from github).

Please fix merge conflicts and I'll merge and release a new version.

fredrike commented:
The tests doesn't work, are you mocking the responses?

fredrike commented on pyproject.toml:
```diff
@@ -8,7 +8,7 @@ authors = [
 ]
 name = "pydaikin"
-version = "2.15.0"
+version = "2.16.0"
```
```suggestion
version = "2.15.0"
```

All version bumping are happening from the GitHub action.

fredrike commented on tests/comprehensive_test.py:
```diff
@@ -0,0 +1,121 @@
+    print("\n=== Testing Daikin Firmware 2.8.0 Integration ===\n")
+
+    # Replace with your device's IP address
+    device_ip = "192.168.50.47"
```
This must be mocked, all test must pass without connecting to external services.

Chris971991 replied:
The comprehensive test was a one off I created. My final test file was the integrated_network_test.py which was what I used to confirm connection to the 2.8.0 firmware device and control it

fredrike replied:
I see, but we require all tests to pass so either configure this to not run automatically or do full mocking of the responses..

https://github.com/fredrike/pydaikin/actions/runs/14875653042/job/41772825413?pr=61#step:9:202

fredrike commented on .bumpversion.cfg:
```diff
@@ -1,5 +1,5 @@
 commit = True
 tag = True
-files = pyproject.toml
+files = pyproject.toml
```
```suggestion
files = pyproject.toml

```

fredrike commented on .bumpversion.cfg:
```diff
@@ -1,5 +1,5 @@
 [bumpversion]
-current_version = 2.15.0
+current_version = 2.16.0
```
```suggestion
current_version = 2.15.0
```

We bump versions in the github action.

fredrike commented on pydaikin/values.py:
```diff
@@ -2,6 +2,9 @@
 from collections.abc import MutableMapping
 from datetime import datetime, timedelta, timezone
+import logging
+
```
Logging is not used in this file.

fredrike commented on pyproject.toml:
```diff
@@ -68,4 +68,4 @@ force_grid_wrap = 0
 indent = "    "
 # will group `import x` and `from x import` of the same module.
-force_sort_within_sections = true
+force_sort_within_sections = true
```
```suggestion
force_sort_within_sections = true

```

We always end with an empty line.

Chris971991 commented:
If interested, this is the also currently working custom intergration into HA - https://github.com/Chris971991/daikin_2_8_0

fredrike commented:
> If interested, this is the also currently working custom intergration into HA - https://github.com/Chris971991/daikin_2_8_0

Nice!

Is this the same as https://github.com/Apoc182/local_daikin?

fredrike commented on README.md:
```diff
@@ -16,11 +16,12 @@ Currently the following Daikin WiFi modules are supported:
 * BRP15B61 aka. AirBase (similar protocol as BRP069Axx)
 * BRP072B/Cxx (needs https access and a key)
 * SKYFi (different protocol, have a password)
+* BRP devices with firmware version 2.8.0 (different API structure) - new in version 2.16.0
```
```suggestion
* BRP084 devices with firmware version 2.8.0 (different API structure)
* SKYFi (different protocol, have a password)
```

fredrike commented on README.md:
```diff
@@ -33,3 +34,13 @@ async def testDaikin():
+
+* FTKM20YVMA with firmware version 2.8.0
+* FTXM46WVMA with firmware version 2.8.0
+* FTXV80WVMA with firmware version 2.8.0
```
```suggestion
* FTXV80WVMA with firmware version 2.8.0
* FTXA25C2V1BW with firmware version 2.8.0
* FTXA50C2V1BW with firmware version 2.8.0
```

According to: https://github.com/home-assistant/core/issues/125584#issuecomment-2854839579

fredrike commented on tests/comprehensive_test.py:
```diff
@@ -0,0 +1,121 @@
+logging.basicConfig(level=logging.DEBUG, format='%(levelname)s: %(message)s')
+
+
+async def test_daikin():
```
```suggestion
@pytest.mark.skip(reason="We don't require connection to real devices")
async def test_daikin():
```

fredrike commented on tests/comprehensive_test.py:
```diff
@@ -0,0 +1,121 @@
+import asyncio
+import logging
```
```suggestion
import logging

import pytest
```

fredrike commented on tests/test_daikin.py:
```diff
@@ -0,0 +1,34 @@
+import asyncio
+import logging
+
```
```suggestion

import pytest

```

fredrike commented on tests/test_daikin.py:
```diff
@@ -0,0 +1,34 @@
+logging.basicConfig(level=logging.DEBUG, format='%(levelname)s: %(message)s')
+
+
+async def test_daikin():
```
```suggestion
@pytest.mark.skip(reason="We don't require connection to real devices")
async def test_daikin():
```

https://docs.pytest.org/en/stable/how-to/skipping.html#skipping-test-functions

Even better would be to read IP from an ENVIRONMENT variable and skip if that env is not set, but this is good enough.

Chris971991 commented:
> > If interested, this is the also currently working custom intergration into HA - https://github.com/Chris971991/daikin_2_8_0
> 
> Nice!
> 
> Is this the same as https://github.com/Apoc182/local_daikin?

Yeah but with some tweaks :) I saw that you were looking to integrate this into pydaikin which was the inspiration for this :D

Chris971991 commented:
This is my ever first pull request, so I may be making mistakes. Please forgive me :(

fredrike commented:
There are still parts in the tests that are failing.

https://github.com/fredrike/pydaikin/actions/runs/14878207521/job/41779822510?pr=61#step:9:935

I'm usually using https://pypi.org/project/pytest-httpx/ for mocking httpx requests.

fredrike commented:
> This is my ever first pull request, so I may be making mistakes. Please forgive me :(

No worries, let's try to get this working together then. Next will be to get approval in HomeAssistant, that have been a great learning experience for me ;).

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+    @staticmethod
+    def temp_to_hex(temperature: float, divisor=2) -> str:
+        """Convert temperature to hexadecimal."""
+        return format(int(temperature * divisor), '02x')
```
Why not use https://docs.python.org/2/library/functions.html#hex?

fredrike replied:
No this is fine.

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+
+
+# pylint: disable=unused-argument
+async def set_zone(self, zone_id, key, value):
```
Make sure that this returns `False` instead.

https://github.com/fredrike/pydaikin/blob/master/pydaikin/daikin_base.py#L354

Chris971991 replied:
Implemented

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+
+
+# pylint: disable=unused-argument
+async def set_advanced_mode(self, mode, value):
```
Make sure this returns `False` instead.

https://github.com/fredrike/pydaikin/blob/master/pydaikin/daikin_base.py#L335

Chris971991 replied:
Implemented

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+
+
+# pylint: disable=unused-argument
+async def set_holiday(self, mode):
```
Make sure this returns `False` instead.

https://github.com/fredrike/pydaikin/blob/master/pydaikin/daikin_base.py#L310

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+            # Get humidity
+            try:
+                self.values['hhum'] = str(
+                    int(
```
Overload hex to int so we don't have this:

`int(<value>, 16)`

Chris971991 replied:
Implemented

fredrike reviewed (changes requested):
I'm wondering if there could be an better way to do the mapping of values. 

Not super found of all these:

```python
self.find_value_by_pn(
                response,
                '/dsiot/edge/adr_0100.dgc_status',
                'dgc_status',
                'e_1002',
                'e_3001',
                'p_01',
            )
```

Perhaps we could create a constant like this:

```
MAPPING = {
  "mode": ['/dsiot/edge/adr_0100.dgc_status', 'dgc_status', 'e_1002', 'e_3001', 'p_01'],
}
```

and then use it like this:

```
mode_value = self.find_value_by_pn(response, *(MAPPING["mode"]))
```
https://github.com/fredrike/pydaikin/pull/61/files#diff-f64931dfa438462bb8227f9002ce52d4660ebeca87c8681b070d3dfd7629df6bR281

It would make the code much cleaer.

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+                    horizontal_attr_name,
+                )
+
+                if horizontal and vertical:
+                    result = 'both'
+                elif horizontal:
+                    result = 'horizontal'
+                elif vertical:
+                    result = 'vertical'
```
```suggestion
                if horizontal and vertical:
                    return "both"
                if horizontal:
                    return "horizontal"
                if vertical:
                    return "vertical"
```

Chris971991 replied:
Implemented in latest pull

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+
+    def get_swing_state(self, data: dict) -> str:
+        """Get the current swing state from response data."""
+        result = 'off'  # Default return value
```
```suggestion
```

Chris971991 replied:
implemented in latest pull

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+                elif vertical:
+                    result = 'vertical'
+            except DaikinException:
+                pass  # Keep default 'off'
```
When does this occur?

fredrike replied:
Ok, let's keep it as it is then.

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,614 @@
+            except DaikinException:
+                pass  # Keep default 'off'
+
+        return result
```
```suggestion
        return "off"
```

Chris971991 replied:
implemented in latest pull

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,617 @@
+    """Set advanced mode. Not supported in this firmware."""
+    _LOGGER.warning("Advanced mode not supported in firmware 2.8.0")
+    # Method intentionally does nothing as this feature is not supported
+    return False
```
Apply similar to all options that aren't supported.

fredrike commented:
> This would clean things up a lot and make future maintenance way easier. What do you think?

Yes, go for it. It will make the rest of the code so much cleaner.

fredrike reviewed (changes requested):
It is not the `set_` method that should return false, it is the `support_` method.

i.e.

https://github.com/fredrike/pydaikin/blob/master/pydaikin/daikin_base.py#L310
https://github.com/fredrike/pydaikin/blob/master/pydaikin/daikin_base.py#L335
https://github.com/fredrike/pydaikin/blob/master/pydaikin/daikin_base.py#L354

fredrike commented on pydaikin/daikin_brp084.py:
```diff
@@ -0,0 +1,629 @@
+
+# pylint: disable=unused-argument
+def set_streamer(self, mode):
+    """Streamer mode not supported in firmware 2.8.0"""
+    _LOGGER.debug("Streamer mode not supported in firmware 2.8.0")
+
+
+# pylint: disable=unused-argument
+async def set_holiday(self, mode):
+    """Set holiday mode."""
+    _LOGGER.debug("Holiday mode not supported in firmware 2.8.0")
+
+
+# pylint: disable=unused-argument
+async def set_advanced_mode(self, mode, value):
+    """Set advanced mode."""
+    _LOGGER.debug("Advanced mode not supported in firmware 2.8.0")
+
+
+async def set_streamer(self, mode):
+    """Set streamer mode."""
+    _LOGGER.debug("Streamer mode not supported in firmware 2.8.0")
+
+
+@property
+def support_away_mode(self) -> bool:
+    """Set holiday mode not supported in firmware 2.8.0"""
+    return False
+
+
+@property
+def support_advanced_modes(self) -> bool:
+    """Advanced mode not supported in firmware 2.8.0"""
+    return False
+
+
+@property
+def support_zone_count(self) -> bool:
+    """Zones mode not supported in firmware 2.8.0"""
+    return False
```
```suggestion

    # pylint: disable=unused-argument
    def set_streamer(self, mode):
        """Streamer mode not supported in firmware 2.8.0"""
        _LOGGER.debug("Streamer mode not supported in firmware 2.8.0")
    
    
    # pylint: disable=unused-argument
    async def set_holiday(self, mode):
        """Set holiday mode."""
        _LOGGER.debug("Holiday mode not supported in firmware 2.8.0")
    
    
    # pylint: disable=unused-argument
    async def set_advanced_mode(self, mode, value):
        """Set advanced mode."""
        _LOGGER.debug("Advanced mode not supported in firmware 2.8.0")
        
    
    @property
    def support_away_mode(self) -> bool:
        """Set holiday mode not supported in firmware 2.8.0"""
        return False
    
    
    @property
    def support_advanced_modes(self) -> bool:
        """Advanced mode not supported in firmware 2.8.0"""
        return False
    
    
    @property
    def support_zone_count(self) -> bool:
        """Zones mode not supported in firmware 2.8.0"""
        return False
```

Chris971991 replied:
Replace missing async

Chris971991 commented:
> > This would clean things up a lot and make future maintenance way easier. What do you think?
> 
> Yes, go for it. It will make the rest of the code so much cleaner.

Implemented, have a look :)

fredrike reviewed (changes requested):
Looks really good, just a few minor suggestions. Perhaps we could create the method `add_request` that does all the list slicing before adding to requests.

i.e.
```python

def add_request(path, value):
    """Append DaikinAttribute to requests."""
    requests.append(DaikinAttribute(path[-1], value, path[2:4], path[0]))
  
```

fredrike commented:
It would be great if there were tests that covers at least the base flow.

Chris971991 commented:
> It would be great if there were tests that covers at least the base flow.

1. Added the add_request helper method

2. Refactored the handler methods:

- _handle_power_setting: Simplified power on/off and mode setting
- _handle_temperature_setting: Streamlined temperature setting
- _handle_fan_setting: Made fan speed setting more concise
- _handle_swing_setting: Greatly simplified vertical and horizontal swing settings

3. Fixed linting issues (self.requests were outside of __init__)

4. Updated testing in test_daikin_brp084.py:

- test_add_request_method: Tests all handler methods using the new add_request method
- test_add_request_direct: Tests the add_request method directly
- All tests are now passing, confirming that the refactored code works correctly.

fredrike commented:
Great job, thanks a million. 

Will you make a pr on home assistant and its docs with the changes necessary for this to work? I’ve created a new version on pypi.

fredrike commented:
> Firmware 3.6.0 just released I found out... HAHA!

What does that mean for this code?

Chris971991 commented:
> > Firmware 3.6.0 just released I found out... HAHA!
> 
> What does that mean for this code?

I am not sure; I will have to look into it. But at least this will resolve the issue for people who upgraded to 2.8.0 and were unaware of the warning to not upgrade. Which will help current users

Chris971991 commented:
> Great job, thanks a million.
> 
> Will you make a pr on home assistant and its docs with the changes necessary for this to work? I’ve created a new version on pypi.

As in updating this? - homeassistant/components/daikin - https://github.com/home-assistant/core/tree/96a8902365ba8e076caa6f4bb8dc323c13114e98/homeassistant/components/daikin

fredrike commented:
> > Great job, thanks a million.
> > Will you make a pr on home assistant and its docs with the changes necessary for this to work? I’ve created a new version on pypi.
> 
> As in updating this? - homeassistant/components/daikin - https://github.com/home-assistant/core/tree/96a8902365ba8e076caa6f4bb8dc323c13114e98/homeassistant/components/daikin

Yes, but I guess that we only need to update the manifest (version bump pydaikin) and then will your changes in factory take care of the rest..

Something like this: https://github.com/home-assistant/core/pull/140424

Also the documentation needs to be updated..

https://www.home-assistant.io/integrations/daikin/

ajvandr commented:
> > > Firmware 3.6.0 just released I found out... HAHA!
> > 
> > 
> > What does that mean for this code?
> 
> I am not sure; I will have to look into it. But at least this will resolve the issue for people who upgraded to 2.8.0 and were unaware of the warning to not upgrade. Which will help current users

I'm one of the guineapigs that tested 2.8.0->3.6.0 using https://github.com/Apoc182/local_daikin and can confirm that there are no apparent changes in the API. Looking forward to seeing support for these units come back to mainstream.
